KAFKA-5610: KafkaApis.HandleWriteTxnMarkerRequest should return UNKNOWN_TOPIC_OR_PARTITION on partition emigration.

apurvam opened the pull request:
Before this patch, we would return the non-retriable `UNSUPPORTED_FOR_MESSAGE_FORMAT` error when leadership changed, causing markers to be lost.

Commit:
Return UKNOWN_TOPIC_OR_PARTITION when partition emigrates during KafkaApis.handleWriteTxnMarkerRequest

apurvam commented:
cc @hachikuji @ijuma

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,15 +1543,19 @@ class KafkaApis(val requestChannel: RequestChannel,
       val producerId = marker.producerId
-      val (goodPartitions, partitionsWithIncorrectMessageFormat) = marker.partitions.asScala.partition { partition =>
+
+      val (goodPartitions, unknownOrIncorrectMessageFormatPartitions) = marker.partitions.asScala.partition { partition =>
```
nit: instead of `goodPartitions`, maybe `validMessageFormatPartitions` or something?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,15 +1543,19 @@ class KafkaApis(val requestChannel: RequestChannel,
       }
 
-      if (partitionsWithIncorrectMessageFormat.nonEmpty) {
+      val (partitionsWithIncorrectMessageFormat, unknownPartitions) = unknownOrIncorrectMessageFormatPartitions.partition(metadataCache.contains(_))
```
Seems this could be moved into the `if` block below?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,15 +1543,19 @@ class KafkaApis(val requestChannel: RequestChannel,
+      if (unknownOrIncorrectMessageFormatPartitions.nonEmpty) {
         val currentErrors = new ConcurrentHashMap[TopicPartition, Errors]()
+        unknownPartitions.foreach { partition => currentErrors.put(partition, Errors.UNKNOWN_TOPIC_OR_PARTITION)}
         partitionsWithIncorrectMessageFormat.foreach { partition => currentErrors.put(partition, Errors.UNSUPPORTED_FOR_MESSAGE_FORMAT) }
```
Discussed offline, but I think we need to ensure that we are only returning this case when we _know_ that the message format version is lower than 2. That means we probably need to handle the case that the partition is in the metadata cache, but we have no local replica (i.e. we have not yet received the LeaderIsr).

hachikuji reviewed (commented):
Good find. Left a few comments.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6161/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6177/
Test PASSed (JDK 7 and Scala 2.11).

Commit:
Revamp the version check logic to be more precise.

apurvam commented:
Thanks for the review. I switched to just using mutable collections for the three cases (correct message format, incorrect message format, unknown message format) and split the input list into each of those three in a single pass. This is more clean IMO and covers all the cases.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6165/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6181/
Test FAILed (JDK 7 and Scala 2.11).

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,24 +1543,34 @@ class KafkaApis(val requestChannel: RequestChannel,
-      val (goodPartitions, partitionsWithIncorrectMessageFormat) = marker.partitions.asScala.partition { partition =>
+      val partitionsWithCorrectMessageFormat = new mutable.ListBuffer[TopicPartition]
+      val partitionsWithIncorrectMessageFormat = new mutable.ListBuffer[TopicPartition]
+      val unknownPartitions = new mutable.ListBuffer[TopicPartition]
```
Nit: `ArrayBuffer` is generally preferred for mutable sequences (similar to `ArrayList` versus `LinkedList` in Java.).

ijuma commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -153,6 +178,53 @@ class KafkaApisTest {
+    EasyMock.expect(metadataCache.contains(tp1))
+      .andReturn(false)
+    EasyMock.expect(metadataCache.contains(tp2))
+      .andReturn(true)
```
Do we care about the metadataCache given the latest update that just relies on `ReplicaManager`? Same for the other tests that include `MetadataCache` assertions.

apurvam replied:
It was required for the previous implementation. Removed them now.

ijuma commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,24 +1543,34 @@ class KafkaApis(val requestChannel: RequestChannel,
+            else
+              partitionsWithCorrectMessageFormat.append(partition)
+          case None =>
+            unknownPartitions.append(partition)
```
Nit: you can use `+=` instead of `append` (e.g. `unknownPartitions += partition`).

ijuma reviewed (commented):
Thanks for the PR, looks good overall, just a few nits.

ijuma commented:
@hachikuji, I assume you are going to cherry-pick this to 0.11.0?

Commit:
Addressed Ismael's comments

hachikuji commented:
@ijuma Yes, I was planning to.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1543,24 +1543,34 @@ class KafkaApis(val requestChannel: RequestChannel,
     for (marker <- markers.asScala) {
       val producerId = marker.producerId
-      val (goodPartitions, partitionsWithIncorrectMessageFormat) = marker.partitions.asScala.partition { partition =>
+      val partitionsWithCorrectMessageFormat = new mutable.ArrayBuffer[TopicPartition]
```
nit: maybe correct/incorect should be compatible/incompatible?

Commit:
Refactor the code to remove two of the mutable collections, per Jason's suggestion

hachikuji reviewed (approved):
Thanks for the patch. LGTM

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6216/
Test PASSed (JDK 7 and Scala 2.11).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6200/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6206/
Test PASSed (JDK 8 and Scala 2.12).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6222/
Test PASSed (JDK 7 and Scala 2.11).

hachikuji commented:
Merging to trunk and 0.11.0.
